Integration/Services/TestReverseProxy.cs

```diff
@@ -0,0 +1,139 @@
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.Extensions.DependencyInjection;
+using System;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using Yarp.ReverseProxy.Forwarder;
+
+namespace Lombiq.Tests.Integration.Services;
+
+public class TestReverseProxy : IDisposable, IAsyncDisposable
+{
+    private bool _disposed;
+    private bool _disposedAsync;
+    private IWebHost _webHost;
+    private IProxyConnectionProvider _proxyConnectionProvider;
+
+    public string RootUrl { get; private set; }
+
+    public TestReverseProxy(string rootUrl) => RootUrl = rootUrl;
+
+    public void AttachConnectionProvider(IProxyConnectionProvider clientConnectionProvider) =>
+        _proxyConnectionProvider = clientConnectionProvider;
+
+    public void DetachConnectionProvider() =>
+        _proxyConnectionProvider = null;
+
+    public Task StartAsync()
+    {
+        if (_webHost != null)
+        {
+            throw new InvalidOperationException("Instance not started.");
+        }
+
+        var webHostBuilder = new WebHostBuilder()
+            .UseKestrel()
+            .UseUrls(RootUrl)
+            .ConfigureServices(services => services
+                .AddHttpForwarder()
+                .AddRouting())
+            .Configure(builder =>
+            {
+                var httpForwarder = builder.ApplicationServices.GetService<IHttpForwarder>();
```

Comment:
Always use `GetRequiredService` unless the service is optional. (`GetService` returns `null` if the service is not found, then you get a hard to debug NRE when you try to use it. If you intentionally want an optional service, then check for null whenever you use it but I don't think that's the case here.)
```suggestion
                var httpForwarder = builder.ApplicationServices.GetRequiredService<IHttpForwarder>();
```